### PR TITLE
Feature/query microservices

### PR DESCRIPTION
**Pull request by jwomeara**

### CODE REVIEW

**Review by jwomeara — COMMENTED** [review]
Notable changes in this PR:
 - FederatedAuthorizationService is the query microservices equivalent of RemoteUserOperations
 - Contains code to dynamically generate FederatedAuthorizationService instances based on yaml properties
 - AuthorizationsListSuppllier added so that microservices can get the correct AuthorizationsListBase implementation
 - DatawaveUserDetails is what we all remember as ProxiedUserDetails from the initial microservices (lots to talk about here)
 - ServerUserDetails supplier is sort of equivalent to the server principal you might see in the webservice and is available to all datawave microservices.

File: src/main/java/datawave/microservice/authorization/federation/DynamicFederatedAuthorizationServiceBeanDefinitionRegistrar.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        federatedAuthorizationProperties.forEach((name, props) -> {
+            GenericBeanDefinition beanDefinition = new GenericBeanDefinition();
+            beanDefinition.setBeanClass(FederatedAuthorizationService.class);
+
+            ConstructorArgumentValues constructorArgValues = new ConstructorArgumentValues();
+            constructorArgValues.addGenericArgumentValue(props);
+            beanDefinition.setConstructorArgumentValues(constructorArgValues);
+
+            beanDefinition.setScope(SCOPE_PROTOTYPE);
+            beanDefinitionRegistry.registerBeanDefinition(name, beanDefinition);
```
Comment (matthpeterson) [inline src/main/java/datawave/microservice/authorization/federation/DynamicFederatedAuthorizationServiceBeanDefinitionRegistrar.java:48]: Looks good.  I went to look at for a test to confirm my understanding and didn't find one.  I wonder if it's worth adding one.

File: src/main/java/datawave/microservice/authorization/federation/config/FederatedAuthorizationServiceProperties.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import javax.validation.constraints.PositiveOrZero;
+
+public class FederatedAuthorizationServiceProperties {
+    private String federatedAuthorizationUri = "https://authorization:8443/authorization/v2";
+    
+    // max bytes to buffer for each rest call (-1 is unlimited)
+    private int maxBytesToBuffer = -1;
+    
+    @Valid
+    private RetryTimeoutProperties listEffectiveAuthorizations = new RetryTimeoutProperties();
```
Comment (matthpeterson) [inline src/main/java/datawave/microservice/authorization/federation/config/FederatedAuthorizationServiceProperties.java:16]: Maybe the field name would be clearer if it had properties in its name.

File: src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                            .build())
+                    .header(ENTITIES_HEADER, getProxiedEntities(currentUser))
+                    .header(ISSUERS_HEADER, getProxiedIssuers(currentUser))
+                    .header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE)
+                    .retrieve()
+                    // don't retry on 4xx errors
+                    .onStatus(HttpStatus::is5xxServerError,
+                            clientResponse -> Mono.error(new ServiceException("Service Error", clientResponse.rawStatusCode())))
+                    .toEntity(GenericResponse.class)
+                    .retryWhen(Retry
```
Comment (matthpeterson) [inline src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java:158]: The argument to retryWhen appears to be identical to what's sent to retryWhen.  Maybe there's some maintenance value in extracting a method for this.

File: src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+                            .fixedDelay(retry.getRetries(), Duration.ofMillis(retry.getRetryDelayMillis()))
+                            .filter(throwable -> throwable instanceof ServiceException)
+                            .onRetryExhaustedThrow(((retryBackoffSpec, retrySignal) -> {
+                                throw new ServiceException("External Service failed to process after max retries",
+                                        HttpStatus.SERVICE_UNAVAILABLE.value());
+                            })))
+                    .block(Duration.ofMillis(retry.getTimeoutMillis()));
+            // @formatter:on
+            
+            AuthorizationException authorizationException;
```
Comment (matthpeterson) [inline src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java:168]: Similarly, it looks like the only difference between this block of code and the other is the name of the action, which could perhaps be a parameter to a new function.

**Review by matthpeterson — APPROVED** [review]

File: src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    
+    private FederatedAuthorizationServiceProperties federatedAuthorizationProperties;
+    private final WebClient webClient;
+    private AuthorizationsListSupplier authorizationsListSupplier;
+    
+    public FederatedAuthorizationService(FederatedAuthorizationServiceProperties federatedAuthorizationProperties, WebClient.Builder webClientBuilder,
+                    AuthorizationsListSupplier authorizationsListSupplier) {
+        this.federatedAuthorizationProperties = federatedAuthorizationProperties;
+        // @formatter:off
+        this.webClient = webClientBuilder
```
Comment (alerman) [inline src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java:43]: Do we set a timeout anywhere, and if not should we be? What happens if the federated auth service times out?

File: src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                        .codecs(clientCodecConfigurer -> clientCodecConfigurer
+                                .defaultCodecs()
+                                .maxInMemorySize(federatedAuthorizationProperties.getMaxBytesToBuffer()))
+                        .build())
+                .build();
+        // @formatter:on
+        this.authorizationsListSupplier = authorizationsListSupplier;
+    }
+    
+    private String getProxiedEntities(ProxiedUserDetails currentUser) {
```
Comment (alerman) [inline src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java:55]: This method and the one below could be collapsed and take as a parameter the field to use (issuer v subject).

File: src/main/java/datawave/microservice/authorization/federation/config/FederatedAuthorizationServiceProperties.java (hunk 2 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/authorization/federation/config/FederatedAuthorizationServiceProperties.java:16]: whatever name you use here is what you will have to write out in the yaml configuration.  my feeling is that adding 'properties' would make this a bit verbose - especially if we did that throughout.

Comment (jwomeara) [reply src/main/java/datawave/microservice/authorization/federation/config/FederatedAuthorizationServiceProperties.java:16]: whoah...  looking at this again and now i see the obvious mistake you were trying to point out.  yikes!  i think i fell victim to the overzealous-ness of intellij's rename feature.

Comment (jwomeara) [reply src/main/java/datawave/microservice/authorization/federation/config/FederatedAuthorizationServiceProperties.java]: changed the naming.  hopefully that's better.

File: src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java (hunk 5 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java:43]: This part is the overall timeout.
```
.block(Duration.ofMillis(retry.getTimeoutMillis()));
```

File: src/main/java/datawave/microservice/authorization/federation/DynamicFederatedAuthorizationServiceBeanDefinitionRegistrar.java (hunk 1 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/authorization/federation/DynamicFederatedAuthorizationServiceBeanDefinitionRegistrar.java:48]: Sure, I can do that.  This stuff is still kind of under development, but that's a good idea.

File: src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java (hunk 3 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java]: should be fixed with the latest push.

File: src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java (hunk 4 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/authorization/federation/FederatedAuthorizationService.java:137]: should be fixed with the latest push.

File: src/main/java/datawave/microservice/authorization/federation/DynamicFederatedAuthorizationServiceBeanDefinitionRegistrar.java (hunk 1 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/authorization/federation/DynamicFederatedAuthorizationServiceBeanDefinitionRegistrar.java:48]: added

**Review by matthpeterson — APPROVED** [review]

**Review by apmoriarty — APPROVED** [review]

**Pull request merged**